debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/EventBuffer.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mysql;
+
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.shyiko.mysql.binlog.event.Event;
+import com.github.shyiko.mysql.binlog.event.EventType;
+import com.github.shyiko.mysql.binlog.event.QueryEventData;
+
+import io.debezium.connector.mysql.BinlogReader.BinlogPosition;
+
+/**
+ * This class represents a look-ahead buffer that allows Debezium to accumulate binlog events and decide
+ * if the last event in transaction is either {@code ROLLBACK} or {@code COMMIT}. The incoming evets are either
+ * supposed to be in transaction or out-of-transaction. When out-of-transaction they are sent directly into
+ * the destination handler. When in transaction the event goes through the buffering.
+ * 
+ * The transaction start is identified by a {@code BEGIN} event. Transaction is ended either by {@code COMMIT}
+ * event or by {@code XID} an event.
+ * 
+ * If there are more events that can fit to the buffer then
+ * <ul>
+ *     <li>Binlog position is recorded for the first event not fitting into the buffer</li>
+ *     <li>Binlog position is recorded for the commit event</li>
+ *     <li>Buffer content is sent to the final handler</li>
+ *     <li>Binlog position is rewound and all events between the above recorded positions are sent to the final handler</li>
+ * </ul>
+ * 
+ * @author Jiri Pechanec
+ *
+ */
+class EventBuffer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EventBuffer.class);
+
+    private final int capacity;
+    private final Queue<Event> buffer;
+    private final BinlogReader reader;
+    private boolean txStarted = false;
+
+    /**
+     * Contains the position of the first event that has not fit into the buffer.
+     */
+    private BinlogPosition largeTxNotBufferedPosition;
+
+    /**
+     * Contains the position of the last event in belonging to the transaction that has not fit into
+     * the buffer.
+     */
+    private BinlogPosition forwardTillPosition;
+
+    public EventBuffer(int capacity, BinlogReader reader) {
+        this.capacity = capacity;
+        this.buffer = new ArrayBlockingQueue<>(capacity);
+        this.reader = reader;
+    }
+
+    /**
+     * An entry point to the buffer that should be used by BinlogReader to push events.
+     * @param event to be stored in the buffer
+     */
+    public void add(Event event) {
+        if (event == null) return;
+        if (isReplayingEventsNotFittedInBuffer(event)) {
```

Comment:
This method is violating a general rule which I'd like us to adhere to: a method should either be a "query method" or a "command method", but not both at the same time. I.e. a method which asks for a result shouldn't perform some mutating action as a side effect. Of course there are exceptions, but in general it helps much with keeping the code comprehensible.

In this case here I'll just extract the call to handleEvent() from the method to this if block itself:

```
if (isReplayingEventsBeyondBufferCapacity()) {
    reader.handleEvent(event);
    return;
}
```